<!doctype html>
<html>
<head>
<title>test_small_mouse_pairseqs_v1_parsed_seqs_probs_mq20_clones</title>

<style>
table {
    font-family: arial, sans-serif;
    border-collapse: collapse;
    width: 100%;
}

td, th {
    border: 1px solid #dddddd;
    text-align: center;
    padding: 8px;
}

tr:nth-child(even) {
    background-color: #dddddd;
}


table.tablesorter {
        font-family:arial;
        background-color: #CDCDCD;
        margin:10px 0pt 15px;
        font-size: 10pt;
        width: 100%;
        text-align: left;
}

table.tablesorter thead tr th, table.tablesorter tfoot tr th {
        background-color: #e6EEEE;
        border: 1px solid #FFF;
        font-size: 10pt;
        padding: 4px;
}

table.tablesorter thead tr .header {
        background-image: url(bg.gif);
        background-repeat: no-repeat;
        background-position: center right;
        cursor: pointer;
}
table.tablesorter tbody td {
        color: #3D3D3D;
        padding: 4px;
        background-color: #FFF;
        vertical-align: top;
}
table.tablesorter tbody tr.odd td {
        background-color:#F0F0F6;
}
table.tablesorter thead tr .headerSortUp {
        background-image: url(asc.gif);
}
table.tablesorter thead tr .headerSortDown {
        background-image: url(desc.gif);
}
table.tablesorter thead tr .headerSortDown, table.tablesorter thead tr .headerSortUp {
background-color: #8dbdd8;
}


</style>


<script type="text/javascript" src="jquery-latest.js"></script>
<script type="text/javascript" src="jquery.tablesorter.js"></script>

</head>
<body>

<h1>Overview:</h1>

This page contains the results of TCR repertoire analyses conducted with the tcr-dist pipeline.
For a general overview of the
analysis methods, please consult the reference(s) mentioned in the "CITING" section of the README.txt file in the
tcr-dist repository. Individual plots should be preceded by some explanatory text below.

<br>
<br>

<h1>Command:</h1>
run_basic_analysis.py --constant_seed --intrasubject_nbrdists --organism mouse --pair_seqs_file ./testing/test_small_mouse_pairseqs_v1.tsv
<br>

<h1>Summary table: (click on headers to sort)</h1>
<table id="summaryTable" style="width:100%" class="tablesorter">
    <caption>Summary information on the dataset</caption>
    <thead>
    <tr>
    <th>epitope</th>
<th>num_individuals</th>
<th>num_parsed_reads</th>
<th>num_clones</th>
<th>clonality<sup>1</sup></th>
<th>TCRdiv-a<sup>2</sup></th>
<th>TCRdiv-b<sup>3</sup></th>
<th>TCRdiv-ab<sup>4</sup></th>
<th>Pshare-a<sup>5</sup></th>
<th>Pshare-b<sup>6</sup></th>
<th>Pshare-ab<sup>7</sup></th>
<th>discrimination (AUROC)<sup>8</sup></th>
<th>heterogeneity (Z)<sup>9</sup></th>

    </tr>
    </thead>
    <tfoot>
<tr><th colspan="13" style="text-align:left"><sup>1</sup>Clonality measured using 1.0-Simpson's Diversity Index, weighted-average across subjects</th></tr>
<tr><th colspan="13" style="text-align:left"><sup>2</sup>Inverse-Simpson's style measure of repertoire diversity using Gaussian-smoothed distance matching (alpha)</th></tr>
<tr><th colspan="13" style="text-align:left"><sup>3</sup>Inverse-Simpson's style measure of repertoire diversity using Gaussian-smoothed distance matching (beta)</th></tr>
<tr><th colspan="13" style="text-align:left"><sup>4</sup>Inverse-Simpson's style measure of repertoire diversity using Gaussian-smoothed distance matching (both chains)</th></tr>
<tr><th colspan="13" style="text-align:left"><sup>5</sup>Probability that a clone from one subject has the same alpha-chain AAs as a clone from another</th></tr>
<tr><th colspan="13" style="text-align:left"><sup>6</sup>Probability that a clone from one subject has the same beta-chain AAs as a clone from another</th></tr>
<tr><th colspan="13" style="text-align:left"><sup>7</sup>Probability that a clone from one subject has the same full-chain AAs as a clone from another</th></tr>
<tr><th colspan="13" style="text-align:left"><sup>8</sup>AUROC for discriminating epitope-specific TCRs from background set using NN-distance score</th></tr>
<tr><th colspan="13" style="text-align:left"><sup>9</sup>Z-score of difference between inter- and intra-subject distances (large and positive suggests subjects may have 'different' repertoires)</th></tr>
</tfoot>
<tbody>
<tr>
<td>F2</td>
<td>3</td>
<td>33</td>
<td>21</td>
<td>0.866</td>
<td>38.8</td>
<td>30.6</td>
<td>113.3</td>
<td>0.000E+00</td>
<td>2.370E-02</td>
<td>0.000E+00</td>
<td>0.880</td>
<td>0.3</td>
</tr>
<tr>
<td>M38</td>
<td>3</td>
<td>45</td>
<td>21</td>
<td>0.848</td>
<td>38.3</td>
<td>3.1</td>
<td>24.4</td>
<td>2.370E-02</td>
<td>2.703E-01</td>
<td>0.000E+00</td>
<td>0.983</td>
<td>0.9</td>
</tr>
<tr>
<td>M45</td>
<td>3</td>
<td>44</td>
<td>40</td>
<td>0.987</td>
<td>20.0</td>
<td>12.6</td>
<td>28.0</td>
<td>1.869E-02</td>
<td>2.899E-02</td>
<td>9.901E-03</td>
<td>0.939</td>
<td>0.1</td>
</tr>

    </tbody>
    </table>
    
<br>

<h1>CDR3 table: (click on headers to sort)</h1>
<table id="cdr3Table" style="width:100%" class="tablesorter">
    <caption>Summary information on the CDR3 loops</caption>
    <thead>
    <tr>
    <th>epitope</th>
<th>N<sup>1</sup></th>
<th>a_len<sup>2</sup></th>
<th>a_charge<sup>3</sup></th>
<th>a_hydro1<sup>4</sup></th>
<th>a_hydro2<sup>5</sup></th>
<th>b_len<sup>6</sup></th>
<th>b_charge<sup>7</sup></th>
<th>b_hydro1<sup>8</sup></th>
<th>b_hydro2<sup>9</sup></th>
<th>ab_len<sup>10</sup></th>
<th>ab_charge<sup>11</sup></th>
<th>ab_hydro1<sup>12</sup></th>
<th>ab_hydro2<sup>13</sup></th>

    </tr>
    </thead>
    <tfoot>
<tr><th colspan="14" style="text-align:left"><sup>1</sup>num_clones</th></tr>
<tr><th colspan="14" style="text-align:left"><sup>2</sup>Mean CDR3-a length</th></tr>
<tr><th colspan="14" style="text-align:left"><sup>3</sup>Mean CDR3-a total charge</th></tr>
<tr><th colspan="14" style="text-align:left"><sup>4</sup>Mean CDR3-a total hydrophobicity (GES scale)</th></tr>
<tr><th colspan="14" style="text-align:left"><sup>5</sup>Mean CDR3-a total hydrophobicity (HP scale)</th></tr>
<tr><th colspan="14" style="text-align:left"><sup>6</sup>Mean CDR3-b length</th></tr>
<tr><th colspan="14" style="text-align:left"><sup>7</sup>Mean CDR3-b total charge</th></tr>
<tr><th colspan="14" style="text-align:left"><sup>8</sup>Mean CDR3-b total hydrophobicity (GES scale)</th></tr>
<tr><th colspan="14" style="text-align:left"><sup>9</sup>Mean CDR3-b total hydrophobicity (HP scale)</th></tr>
<tr><th colspan="14" style="text-align:left"><sup>10</sup>Mean CDR3-ab length</th></tr>
<tr><th colspan="14" style="text-align:left"><sup>11</sup>Mean CDR3-ab total charge</th></tr>
<tr><th colspan="14" style="text-align:left"><sup>12</sup>Mean CDR3-ab total hydrophobicity (GES scale)</th></tr>
<tr><th colspan="14" style="text-align:left"><sup>13</sup>Mean CDR3-ab total hydrophobicity (HP scale)</th></tr>
</tfoot>
<tbody>
<tr>
<td>F2</td>
<td>21</td>
<td>13.0</td>
<td>0.3</td>
<td>-7.0</td>
<td>-0.73</td>
<td>13.6</td>
<td>-1.1</td>
<td>-3.6</td>
<td>-0.75</td>
<td>26.7</td>
<td>-0.8</td>
<td>-10.6</td>
<td>-1.48</td>
</tr>
<tr>
<td>M38</td>
<td>21</td>
<td>14.3</td>
<td>0.7</td>
<td>2.0</td>
<td>0.40</td>
<td>16.7</td>
<td>-0.9</td>
<td>0.3</td>
<td>-0.48</td>
<td>31.0</td>
<td>-0.2</td>
<td>2.3</td>
<td>-0.08</td>
</tr>
<tr>
<td>M45</td>
<td>40</td>
<td>13.5</td>
<td>0.4</td>
<td>-1.2</td>
<td>-0.45</td>
<td>14.2</td>
<td>-0.6</td>
<td>-4.1</td>
<td>-1.09</td>
<td>27.7</td>
<td>-0.1</td>
<td>-5.3</td>
<td>-1.54</td>
</tr>

    </tbody>
    </table>
    
<br>


<script>
	$(document).ready(function()
  {
  $("#summaryTable").tablesorter();
  $("#cdr3Table").tablesorter();
  }
	);

	// add parser through the tablesorter addParser method
	$.tablesorter.addParser({
	// set a unique id
	id: 'scinot',
	is: function(s) {
  return /[+\-]?(?:0|[1-9]\d*)(?:\.\d*)?(?:[eE][+\-]?\d+)?/.test(s);
	},
	format: function(s) {
  return $.tablesorter.formatFloat(s);
	},
	type: 'numeric'
	});

</script>




<h1>TCR Clustering trees</h1>
The following links point to pages containing detailed TCRdist hierarchical clustering trees,
grouped by epitope (the first set) or coloring scheme (the second set).
<br>
<br>
"clonality" trees are colored by the size of each TCR clonotype (number of members).
<br>
<br>
"cross_reactivity" trees are colored by occurrence of TCRs in other epitope datasets.
<br>
<br>
"min_other_nbrdist" trees are colored by the lowest (percentiled) nbrdist score for an epitope
other than the source epitope for each receptor, to get an idea of which other repertoires a
given receptor is similar to.
<br>
<br>
"probs" trees are colored by TCR generation probability according to a very simple model of the
rearrangement process.
<br>
<br>
"sharing" trees are colored by the number of subjects a given TCR occurs in.
<br>
<br>
The trees are annotated to the left with the subject (truncated but prefixed with a unique integer), CDR3 amino acid
sequence, non-germline CDR3 amino acids, number of insertions and deletions, clone size, number of subjects
in which the TCR occurs (same repertoire, all repertoires), number of repertoires in which the TCR occurs,
gene families, and the numerical value of the coloring parameter (and any associated string-tag where relevant) if
not already present in the preceding info.
<br>
<br>
<br>
F2 trees A <a href="trees_F2_A_png.html">(png)</a> <a href="trees_F2_A_svg.html">(svg)</a><br>
F2 trees B <a href="trees_F2_B_png.html">(png)</a> <a href="trees_F2_B_svg.html">(svg)</a><br>
F2 trees AB <a href="trees_F2_AB_png.html">(png)</a> <a href="trees_F2_AB_svg.html">(svg)</a><br>
M38 trees A <a href="trees_M38_A_png.html">(png)</a> <a href="trees_M38_A_svg.html">(svg)</a><br>
M38 trees B <a href="trees_M38_B_png.html">(png)</a> <a href="trees_M38_B_svg.html">(svg)</a><br>
M38 trees AB <a href="trees_M38_AB_png.html">(png)</a> <a href="trees_M38_AB_svg.html">(svg)</a><br>
M45 trees A <a href="trees_M45_A_png.html">(png)</a> <a href="trees_M45_A_svg.html">(svg)</a><br>
M45 trees B <a href="trees_M45_B_png.html">(png)</a> <a href="trees_M45_B_svg.html">(svg)</a><br>
M45 trees AB <a href="trees_M45_AB_png.html">(png)</a> <a href="trees_M45_AB_svg.html">(svg)</a><br>
clonality trees A <a href="trees_clonality_A_png.html">(png)</a> <a href="trees_clonality_A_svg.html">(svg)</a><br>
clonality trees B <a href="trees_clonality_B_png.html">(png)</a> <a href="trees_clonality_B_svg.html">(svg)</a><br>
clonality trees AB <a href="trees_clonality_AB_png.html">(png)</a> <a href="trees_clonality_AB_svg.html">(svg)</a><br>
cross_reactivity trees A <a href="trees_cross_reactivity_A_png.html">(png)</a> <a href="trees_cross_reactivity_A_svg.html">(svg)</a><br>
cross_reactivity trees B <a href="trees_cross_reactivity_B_png.html">(png)</a> <a href="trees_cross_reactivity_B_svg.html">(svg)</a><br>
cross_reactivity trees AB <a href="trees_cross_reactivity_AB_png.html">(png)</a> <a href="trees_cross_reactivity_AB_svg.html">(svg)</a><br>
min_other_nbrdist trees A <a href="trees_min_other_nbrdist_A_png.html">(png)</a> <a href="trees_min_other_nbrdist_A_svg.html">(svg)</a><br>
min_other_nbrdist trees B <a href="trees_min_other_nbrdist_B_png.html">(png)</a> <a href="trees_min_other_nbrdist_B_svg.html">(svg)</a><br>
min_other_nbrdist trees AB <a href="trees_min_other_nbrdist_AB_png.html">(png)</a> <a href="trees_min_other_nbrdist_AB_svg.html">(svg)</a><br>
probs trees A <a href="trees_probs_A_png.html">(png)</a> <a href="trees_probs_A_svg.html">(svg)</a><br>
probs trees B <a href="trees_probs_B_png.html">(png)</a> <a href="trees_probs_B_svg.html">(svg)</a><br>
probs trees AB <a href="trees_probs_AB_png.html">(png)</a> <a href="trees_probs_AB_svg.html">(svg)</a><br>
sharing trees A <a href="trees_sharing_A_png.html">(png)</a> <a href="trees_sharing_A_svg.html">(svg)</a><br>
sharing trees B <a href="trees_sharing_B_png.html">(png)</a> <a href="trees_sharing_B_svg.html">(svg)</a><br>
sharing trees AB <a href="trees_sharing_AB_png.html">(png)</a> <a href="trees_sharing_AB_svg.html">(svg)</a><br>
<br><br><h1>_cdr3_distributions.png</h1>

<u>Command</u>:
/shared/cs_researcher/Bradley_P/BradleyLab/pbradley/dat/tcr-dist/make_summary_table.py --clones_file ./testing/test_small_mouse_pairseqs_v1_parsed_seqs_probs_mq20_clones.tsv
<br><br>
<u>Filename</u>:
./testing/test_small_mouse_pairseqs_v1_parsed_seqs_probs_mq20_clones_cdr3_distributions.png
<br><br>
<u>Readme</u>:

Distributions of CDR3 properties for the different epitopes
<br><br>
<br>
<img src="test_small_mouse_pairseqs_v1_parsed_seqs_probs_mq20_clones_cdr3_distributions.png" />
<br>
<br><br><h1>_kpca.png</h1>

<u>Command</u>:
/shared/cs_researcher/Bradley_P/BradleyLab/pbradley/dat/tcr-dist/make_kpca_plots.py --organism mouse --clones_file ./testing/test_small_mouse_pairseqs_v1_parsed_seqs_probs_mq20_clones.tsv --showmotifs
<br><br>
<u>Filename</u>:
./testing/test_small_mouse_pairseqs_v1_parsed_seqs_probs_mq20_clones_kpca.png
<br><br>
<u>Readme</u>:

Kernel Principle Components Analysis (kPCA) 2D projection plots for the repertoires. Each row is a repertoire,
and each point in the plots corresponds to a single TCR clone, with the points arranged so as to keep nearby
TCRs (as measured by TCRdist) nearby in 2 dimensions. The four different panels are the same 2D projection
colored by gene usage for the four different segments (left to right: Va,Ja,Vb,Jb)
<br><br>
<br>
<img src="test_small_mouse_pairseqs_v1_parsed_seqs_probs_mq20_clones_kpca.png" />
<br>
<br><br><h1>_sharing_diversity.png</h1>

<u>Command</u>:
/shared/cs_researcher/Bradley_P/BradleyLab/pbradley/dat/tcr-dist/plot_sharing.py --organism mouse --clones_file ./testing/test_small_mouse_pairseqs_v1_parsed_seqs_probs_mq20_clones.tsv
<br><br>
<u>Filename</u>:
./testing/test_small_mouse_pairseqs_v1_parsed_seqs_probs_mq20_clones_sharing_diversity.png
<br><br>
<u>Readme</u>:
These bar plots represent three ways of looking at the total diversity of an epitope-specific
repertoire, measured over single-chains (top two rows) and full receptors (bottom row). The left column shows the TCRdiv
diversity measure (a variant of the (inverse of) Simpson's Diversity Index that accounts for receptor similarity
as well as identity). The middle column shows the average NNdistance aka nbrdistance over all the TCR clones in the
repertoire and captures the average distance from TCR clones to their nearest neighbors in the repertoire.
The right column is based on comparison between the epitope-specific TCRs and random TCRs, specifically
looking at the "AUROC" measure. AUROC measures the area under the ROC curve, and captures the trade-off between true and
false positives as the value of a quality or sorting metric is varied from best to worst. An AUROC of 100 (perfect
separation) means that all the true positives occur before any of the false positives. An AUROC of 50 (random)
means that the positives and negatives are equally distributed with respect to the sorting measure. The sort used here
is the "nbrdist10P" which measures the weighted average distance to the closest tenth of the reference repertoire.
The AUROCs
are plotted with a flipped y-axis for easy comparison to the other columns,
the idea being that a more diverse repertoire will be harder to reliably separate
from the random background and hence will have smaller AUROC values (ie, lower AUROC == higher diversity).


<br><br>
<br>
<img src="test_small_mouse_pairseqs_v1_parsed_seqs_probs_mq20_clones_sharing_diversity.png" />
<br>
<br><br><h1>_subject_heterogeneity.png</h1>

<u>Command</u>:
/shared/cs_researcher/Bradley_P/BradleyLab/pbradley/dat/tcr-dist/analyze_epitope_privacy.py --constant_seed --organism mouse --clones_file ./testing/test_small_mouse_pairseqs_v1_parsed_seqs_probs_mq20_clones.tsv --all_chains AB --nrepeat 1000 --tree_height_inches 5.0 --nbrdist_percentile 10
<br><br>
<u>Filename</u>:
./testing/test_small_mouse_pairseqs_v1_parsed_seqs_probs_mq20_clones_subject_heterogeneity.png
<br><br>
<u>Readme</u>:
This summary plot is aimed at answering the question: are some mice (or humans) sampling from a different repertoire than others?
There is a companion plot later on (_subject_trees.png) which adds a little more detail. The idea behind the analysis is to compute
distances between all TCRs in the dataset, and then split those distances into two sets: intra-mouse distances (distances between
TCRs from the same mouse), and inter-mouse distances (distances between TCRs from different mice). If mice are sampling from different
repertoires then the intra-mouse distances might be expected to be smaller than the inter-mouse distances. To assess significance of an
observed difference, I use randomization: the TCRs are randomly re-assigned to the mice 1000 times (preserving the size of each mouse's
individual repertoire), and each time the intra- and inter-mouse
distances are calculated and the mean values of each are stored. To assess the significance of the actually observed mean intra- and inter-
mouse distances, I compute the mean and standard deviation of the 1000 shuffled means and use those to assign a Z-score to the observed
values. Larger absolute values of the Z-score suggest greater significance. To get an approximate P-value I keep track of how often the
random mean intra-mouse distance is less than the observed mean intra-mouse distance, and divide that number by 1000 (the number of random
reshufflings). This bar plot shows the Z-scores; the P-values are listed below the epitopes at the bottom.

<br><br>
<br>
<img src="test_small_mouse_pairseqs_v1_parsed_seqs_probs_mq20_clones_subject_heterogeneity.png" />
<br>
<br><br><h1>_subject_table.png</h1>

<u>Command</u>:
/shared/cs_researcher/Bradley_P/BradleyLab/pbradley/dat/tcr-dist/make_mouse_table.py --clones_file ./testing/test_small_mouse_pairseqs_v1_parsed_seqs_probs_mq20_clones.tsv
<br><br>
<u>Filename</u>:
./testing/test_small_mouse_pairseqs_v1_parsed_seqs_probs_mq20_clones_subject_table.png
<br><br>
<u>Readme</u>:
This subject-table plot shows all the successfully parsed, paired reads, split by mouse/subject (the rows)
and epitope (the columns, labeled at the bottom). The epitope column labels include in parentheses the number of clones followed by
the total number of TCRs. Each pie shows the paired reads for a single mouse/epitope combination, with each wedge corresponding to
a clone. The size of the top clone is shown in red near the red wedge, and the total number of reads is shown below the pie in black.

<br><br>
<br>
<img src="test_small_mouse_pairseqs_v1_parsed_seqs_probs_mq20_clones_subject_table.png" />
<br>
<br><br><h1>_gene_segment_pies.png</h1>

<u>Command</u>:
/shared/cs_researcher/Bradley_P/BradleyLab/pbradley/dat/tcr-dist/make_gene_plots.py --organism mouse --clones_file ./testing/test_small_mouse_pairseqs_v1_parsed_seqs_probs_mq20_clones.tsv --use_color_gradients
<br><br>
<u>Filename</u>:
./testing/test_small_mouse_pairseqs_v1_parsed_seqs_probs_mq20_clones_gene_segment_pies.png
<br><br>
<u>Readme</u>:
These pie charts depict the gene segment composition of the epitope-specific datasets,
with each column corresponding to an epitope (labeled at the bottom, with total number of clones given in
parentheses). Genes making up at least 5 percent of the population are labeled. The rows correspond to
V-alpha, J-alpha, V-beta, and J-beta distributions. At the bottom, the epitopes are clustered based on these gene-frequencies
using the Jensen-Shannon divergence measure.

<br><br>
<br>
<img src="test_small_mouse_pairseqs_v1_parsed_seqs_probs_mq20_clones_gene_segment_pies.png" />
<br>
<br><br><h1>_cdr3lens.png</h1>

<u>Command</u>:
/shared/cs_researcher/Bradley_P/BradleyLab/pbradley/dat/tcr-dist/make_gene_plots.py --organism mouse --clones_file ./testing/test_small_mouse_pairseqs_v1_parsed_seqs_probs_mq20_clones.tsv --use_color_gradients
<br><br>
<u>Filename</u>:
./testing/test_small_mouse_pairseqs_v1_parsed_seqs_probs_mq20_clones_cdr3lens.png
<br><br>
<u>Readme</u>:
These bar plots show the cdr3-length distributions for each epitope, colored by
gene segment. Each epitope is a single row. The left two columns show CDR3-alpha length distributions,
colored by V-segment in the first column and J-segment in the second column. The right two columns depict the
CDR3-beta length distributions in the same manner. Segments comprising at least 5 percent of the epitope's
dataset are labeled.

<br><br>
<br>
<img src="test_small_mouse_pairseqs_v1_parsed_seqs_probs_mq20_clones_cdr3lens.png" />
<br>
<br><br><h1>_gene_entropies_and_mi.png</h1>

<u>Command</u>:
/shared/cs_researcher/Bradley_P/BradleyLab/pbradley/dat/tcr-dist/make_gene_plots.py --organism mouse --clones_file ./testing/test_small_mouse_pairseqs_v1_parsed_seqs_probs_mq20_clones.tsv --use_color_gradients
<br><br>
<u>Filename</u>:
./testing/test_small_mouse_pairseqs_v1_parsed_seqs_probs_mq20_clones_gene_entropies_and_mi.png
<br><br>
<u>Readme</u>:
These three plots contain information on the gene-segment distributions and
correlations. <br><br>
The top plot (gene entropies) shows the total Shannon entropy of the gene segment
probability distributions for each epitope (columns) and the four gene types (rows: VA,JA,VB,JB). Red (higher
entropy) means more disorder (total frequency divided more evenly among larger numbers of genes) while
blue (less entropy) reflects stronger preference for one or a few genes. Segments like V-alpha with a lot of
possible genes will tend to have higher entropies than segments like J-beta with only a few.
<br><br>
The middle plot (gene relative entropies) shows the difference in gene frequency distributions between the observed, epitope-specific
repertoires and control or background distributions derived from large, non-epitope-specific datasets. The number being plotted
is the Jensen-Shannon divergence between the background and observed distributions, divided by the mean Shannon entropy
of the background and observed distributions.
<br><br>
The bottom plot (gene-gene adjusted mutual information) shows the mutual information between the different
gene segment distributions, with red indicating that knowledge of the identity of one gene (for example V-alpha) gives
more information about the identity of another (e.g., V-beta), and blue suggesting that the corresponding gene-pair
for that row are approximately independent for the epitope corresponding to that column. The actual number ranges
from 0 (completely independent) to 1 (completely dependent) and is an adjusted form of the mutual information that
takes into account the numbers of counts (computed using the scikit-learn routine sklearn.metrics.adjusted_mutual_info_score)

<br><br>
<br>
<img src="test_small_mouse_pairseqs_v1_parsed_seqs_probs_mq20_clones_gene_entropies_and_mi.png" />
<br>
<br><br><h1>_nbrdist10_distributions_w_cdf.png</h1>

<u>Command</u>:
/shared/cs_researcher/Bradley_P/BradleyLab/pbradley/dat/tcr-dist/plot_nbrdist_distributions.py --clones_file ./testing/test_small_mouse_pairseqs_v1_parsed_seqs_probs_mq20_clones.tsv --nbrdist_percentiles 5 10 25
<br><br>
<u>Filename</u>:
./testing/test_small_mouse_pairseqs_v1_parsed_seqs_probs_mq20_clones_nbrdist10_distributions_w_cdf.png
<br><br>
<u>Readme</u>:
These plots show distributions of the nbr-distance (aka NNdistance) score for the different
epitopes in the dataset. The score assigned to a given TCR is just the weighted average of the distance
to the nearest 10th percentile
of the other TCRs in the repertoire. Lower values indicate more clustering nearby the TCR in question. The three
columns are based on (left to right) alpha-chain-only, beta-chain-only, and paired-chains distances.
The top row shows the smoothed frequency distributions of nbr-distance score, and the bottom row
shows the cumulative distributions: the height of the curve above a given nbrdistance value is the fraction
of the repertoire with scores less than or equal to that value (so a more clustered repertoire will have a curve that rises
higher earlier than a less clustered repertoire). The legends give the average value of
nbr-distance over the entire repertoire. Lower values correspond to more clustered / less diverse repertoires. These
averages are also the areas above the CDF curves, and are the diversity metrics plotted in the middle panels of the diversity
summary plots.
    
<br><br>
<br>
<img src="test_small_mouse_pairseqs_v1_parsed_seqs_probs_mq20_clones_nbrdist10_distributions_w_cdf.png" />
<br>
<br><br><h1>_nbrdist25_distributions_w_cdf.png</h1>

<u>Command</u>:
/shared/cs_researcher/Bradley_P/BradleyLab/pbradley/dat/tcr-dist/plot_nbrdist_distributions.py --clones_file ./testing/test_small_mouse_pairseqs_v1_parsed_seqs_probs_mq20_clones.tsv --nbrdist_percentiles 5 10 25
<br><br>
<u>Filename</u>:
./testing/test_small_mouse_pairseqs_v1_parsed_seqs_probs_mq20_clones_nbrdist25_distributions_w_cdf.png
<br><br>
<u>Readme</u>:
These plots show distributions of the nbr-distance (aka NNdistance) score for the different
epitopes in the dataset. The score assigned to a given TCR is just the weighted average of the distance
to the nearest 25th percentile
of the other TCRs in the repertoire. Lower values indicate more clustering nearby the TCR in question. The three
columns are based on (left to right) alpha-chain-only, beta-chain-only, and paired-chains distances.
The top row shows the smoothed frequency distributions of nbr-distance score, and the bottom row
shows the cumulative distributions: the height of the curve above a given nbrdistance value is the fraction
of the repertoire with scores less than or equal to that value (so a more clustered repertoire will have a curve that rises
higher earlier than a less clustered repertoire). The legends give the average value of
nbr-distance over the entire repertoire. Lower values correspond to more clustered / less diverse repertoires. These
averages are also the areas above the CDF curves, and are the diversity metrics plotted in the middle panels of the diversity
summary plots.
    
<br><br>
<br>
<img src="test_small_mouse_pairseqs_v1_parsed_seqs_probs_mq20_clones_nbrdist25_distributions_w_cdf.png" />
<br>
<br><br><h1>_random_nbrdists_nbrdists_AB.png</h1>

<u>Command</u>:
/shared/cs_researcher/Bradley_P/BradleyLab/pbradley/dat/tcr-dist/read_random_tcr_distances.py --organism mouse --clones_file ./testing/test_small_mouse_pairseqs_v1_parsed_seqs_probs_mq20_clones.tsv
<br><br>
<u>Filename</u>:
./testing/test_small_mouse_pairseqs_v1_parsed_seqs_probs_mq20_clones_random_nbrdists_nbrdists_AB.png
<br><br>
<u>Readme</u>:
KDE-smoothed histograms of different AB nbrdist measures, comparing each epitope-specific repertoire (red)
            to TCRs from the other repertoires (green) and to random TCRs (blue).<br><br>
            
<br><br>
<br>
<img src="test_small_mouse_pairseqs_v1_parsed_seqs_probs_mq20_clones_random_nbrdists_nbrdists_AB.png" />
<br>
<br><br><h1>_random_nbrdists_roc_AB.png</h1>

<u>Command</u>:
/shared/cs_researcher/Bradley_P/BradleyLab/pbradley/dat/tcr-dist/read_random_tcr_distances.py --organism mouse --clones_file ./testing/test_small_mouse_pairseqs_v1_parsed_seqs_probs_mq20_clones.tsv
<br><br>
<u>Filename</u>:
./testing/test_small_mouse_pairseqs_v1_parsed_seqs_probs_mq20_clones_random_nbrdists_roc_AB.png
<br><br>
<u>Readme</u>:
ROC curves showing true- (y-axis) versus false-positives (x-axis) as the sorting metric increases. Legends give the area under the curve (AUROC) values, ranging from 100 (perfect discrimination, curve goes straight
            up then straight across), to 50 (random), to 0 (total failure, all negatives come before all positives, curve goes straight across then straight up).<br><br>
            
<br><br>
<br>
<img src="test_small_mouse_pairseqs_v1_parsed_seqs_probs_mq20_clones_random_nbrdists_roc_AB.png" />
<br>
<br><br><h1>_random_nbrdists_summary.png</h1>

<u>Command</u>:
/shared/cs_researcher/Bradley_P/BradleyLab/pbradley/dat/tcr-dist/read_random_tcr_distances.py --organism mouse --clones_file ./testing/test_small_mouse_pairseqs_v1_parsed_seqs_probs_mq20_clones.tsv
<br><br>
<u>Filename</u>:
./testing/test_small_mouse_pairseqs_v1_parsed_seqs_probs_mq20_clones_random_nbrdists_summary.png
<br><br>
<u>Readme</u>:
KDE-smoothed nbrdist10P histograms for alpha (col 1), beta (col 2), and alpha-beta (col 3). ROC curves for
the same in columns 4-6.<br><br>
<br><br>
<br>
<img src="test_small_mouse_pairseqs_v1_parsed_seqs_probs_mq20_clones_random_nbrdists_summary.png" />
<br>
<br><br><h1>_random_nbrdists_nbrdist_roc_superpositions.png</h1>

<u>Command</u>:
/shared/cs_researcher/Bradley_P/BradleyLab/pbradley/dat/tcr-dist/read_random_tcr_distances.py --organism mouse --clones_file ./testing/test_small_mouse_pairseqs_v1_parsed_seqs_probs_mq20_clones.tsv
<br><br>
<u>Filename</u>:
./testing/test_small_mouse_pairseqs_v1_parsed_seqs_probs_mq20_clones_random_nbrdists_nbrdist_roc_superpositions.png
<br><br>
<u>Readme</u>:

Superimposed KDE-smoothed NNdistance (wtd_nbrdist10) distributions (left) and ROC curves (right) for all epitopes (paired-chain analyses).<br><br>

<br><br>
<br>
<img src="test_small_mouse_pairseqs_v1_parsed_seqs_probs_mq20_clones_random_nbrdists_nbrdist_roc_superpositions.png" />
<br>
<br><br><h1>_sharing_and_clonality_wtd_nbrdist10.png</h1>

<u>Command</u>:
/shared/cs_researcher/Bradley_P/BradleyLab/pbradley/dat/tcr-dist/analyze_overlap_compute_simpsons.py --organism mouse --clones_file ./testing/test_small_mouse_pairseqs_v1_parsed_seqs_probs_mq20_clones.tsv
<br><br>
<u>Filename</u>:
./testing/test_small_mouse_pairseqs_v1_parsed_seqs_probs_mq20_clones_sharing_and_clonality_wtd_nbrdist10.png
<br><br>
<u>Readme</u>:
These plots explore the relationship between clonality and sharing of TCRs across mice for the same epitope. For the purpose of
    this analysis a TCR is "clonal" if it has a clone_size of at least 2 and is "shared" if it is seen in more than one mouse (ie subject). protprob and nucprob are the
    amino acid and nucleotide (respectively) generation probabilities under a very simple model of the rearrangement process. nbrdist_rank_score is a measure of
    repertoire sampling density nearby a given TCR: we compute an average distance to a TCRs nearest neighbors ("_wtd_nbrdist10rank") and then percentile this over the repertoire to
    get a normalized nbr-distance measure that goes from 0 (many nearby TCRs in the repertoire) to 100 (very few).
    
<br><br>
<br>
<img src="test_small_mouse_pairseqs_v1_parsed_seqs_probs_mq20_clones_sharing_and_clonality_wtd_nbrdist10.png" />
<br>
<br><br><h1>_sharing_and_clonality_by_epitope_wtd_nbrdist10.png</h1>

<u>Command</u>:
/shared/cs_researcher/Bradley_P/BradleyLab/pbradley/dat/tcr-dist/analyze_overlap_compute_simpsons.py --organism mouse --clones_file ./testing/test_small_mouse_pairseqs_v1_parsed_seqs_probs_mq20_clones.tsv
<br><br>
<u>Filename</u>:
./testing/test_small_mouse_pairseqs_v1_parsed_seqs_probs_mq20_clones_sharing_and_clonality_by_epitope_wtd_nbrdist10.png
<br><br>
<u>Readme</u>:
Same plots as above, but now broken down by epitope.
    
<br><br>
<br>
<img src="test_small_mouse_pairseqs_v1_parsed_seqs_probs_mq20_clones_sharing_and_clonality_by_epitope_wtd_nbrdist10.png" />
<br>
<br><br><h1>_epitope_correlations_10.png</h1>

<u>Command</u>:
/shared/cs_researcher/Bradley_P/BradleyLab/pbradley/dat/tcr-dist/plot_nbrdist_distributions.py --clones_file ./testing/test_small_mouse_pairseqs_v1_parsed_seqs_probs_mq20_clones.tsv --nbrdist_percentiles 5 10 25
<br><br>
<u>Filename</u>:
./testing/test_small_mouse_pairseqs_v1_parsed_seqs_probs_mq20_clones_epitope_correlations_10.png
<br><br>
<u>Readme</u>:
These heat maps show correlations between nbrdist scores for different repertoire datasets.
The idea is that, given one reference epitope-specific repertoire, we can assign a number to any TCR in any of the
different epitope-specific repertoires which is the average distance
to the nearest TCRs in the reference repertoire. Two different repertoires give us two different sets of numbers that we
can compare
over the entire merged set of TCRs and ask how well they correlate. Two very similar repertoires should give similar
nbrdist scores for most TCRs in the big dataset, whereas very different repertoires will have uncorrelated nbrdist
scores. What's plotted is just the linear Pearson correlation coefficient between the nbrdist scores assigned by
each pair of reference repertoires, colored from blue (values less than 0.4) to dark red (1.0 -- perfect correlation).

<br><br>
<br>
<img src="test_small_mouse_pairseqs_v1_parsed_seqs_probs_mq20_clones_epitope_correlations_10.png" />
<br>
<br><br><h1>_epitope_epitope_avg_nbrdist_rank_scores.png</h1>

<u>Command</u>:
/shared/cs_researcher/Bradley_P/BradleyLab/pbradley/dat/tcr-dist/plot_nbrdist_distributions.py --clones_file ./testing/test_small_mouse_pairseqs_v1_parsed_seqs_probs_mq20_clones.tsv --nbrdist_percentiles 5 10 25
<br><br>
<u>Filename</u>:
./testing/test_small_mouse_pairseqs_v1_parsed_seqs_probs_mq20_clones_epitope_epitope_avg_nbrdist_rank_scores.png
<br><br>
<u>Readme</u>:
This heat map shows a distance measure between repertoires based on the NNdistance rank score: for any pair of repertoires, we can compute
the NNdistance rank scores of the TCRs in the first repertoire with respect to the second as a reference, and also the NNdistance rank scores of the TCRs in the second repertoire
now using the first as the reference. We can average these numbers to get an average 'mutual' NNdistance rank score for a pair of repertoires. Lower means closer, and
is colored red here.

<br><br>
<br>
<img src="test_small_mouse_pairseqs_v1_parsed_seqs_probs_mq20_clones_epitope_epitope_avg_nbrdist_rank_scores.png" />
<br>
<br><br><h1>_epitope_distances.png</h1>

<u>Command</u>:
/shared/cs_researcher/Bradley_P/BradleyLab/pbradley/dat/tcr-dist/plot_nbrdist_distributions.py --clones_file ./testing/test_small_mouse_pairseqs_v1_parsed_seqs_probs_mq20_clones.tsv --nbrdist_percentiles 5 10 25
<br><br>
<u>Filename</u>:
./testing/test_small_mouse_pairseqs_v1_parsed_seqs_probs_mq20_clones_epitope_distances.png
<br><br>
<u>Readme</u>:
Another way of comparing two repertoires/datasets is just to look at all the distances between a TCR in
one repertoire and a TCR in another, If they are similar repertoires, these numbers will be smaller than if the
repertoires are very different. What's plotted here is the average of the top 10 percent of inter-repertoire TCR distances,
with a color scale running from red (low distances, similar repertoires) to blue (high distances, divergent repertoires).

<br><br>
<br>
<img src="test_small_mouse_pairseqs_v1_parsed_seqs_probs_mq20_clones_epitope_distances.png" />
<br>
<br><br><h1>_sharing.png</h1>

<u>Command</u>:
/shared/cs_researcher/Bradley_P/BradleyLab/pbradley/dat/tcr-dist/plot_sharing.py --organism mouse --clones_file ./testing/test_small_mouse_pairseqs_v1_parsed_seqs_probs_mq20_clones.tsv
<br><br>
<u>Filename</u>:
./testing/test_small_mouse_pairseqs_v1_parsed_seqs_probs_mq20_clones_sharing.png
<br><br>
<u>Readme</u>:
These next five plots reflect different notions of sharing or repetition in the repertoire (ie, seeing "the same" TCR more than once).
The top two plots corresponds to repetition within a single mouse (ie, clonality), and give two representations of Simpson's measure: 1-Simpson's in the top
plot and 1/Simpson's in the bottom. The top plot shows 1-Simpson's for each of the individual mice (labeled by #reads in the mouse) together with an averaged value over
the entire repertoire (red disk, mice are weighted based on number of reads).
<br>
The middle plot uses the Simpson's diversity framework to analyze sharing across mice for the same epitope. The three colors correspond to three notions of
sharing: red points-- seeing the exact same TCR (alpha or beta or both chains); blue points-- seeing two TCRs within a (small) distance of one another; green points-- a
Gaussian-smoothed version of the blue points ("TCRdiv", see explanatory text for the very first figure on this page).
<br>
The bottom two plots look at cross-reactivity: sharing of TCRs between epitopes, either between different mice (plot#4) or within the same mouse (plot#5). The colors
are the same as in the middle plot.

<br><br>
<br>
<img src="test_small_mouse_pairseqs_v1_parsed_seqs_probs_mq20_clones_sharing.png" />
<br>
<br><br><h1>_subject_trees.png</h1>

<u>Command</u>:
/shared/cs_researcher/Bradley_P/BradleyLab/pbradley/dat/tcr-dist/analyze_epitope_privacy.py --constant_seed --organism mouse --clones_file ./testing/test_small_mouse_pairseqs_v1_parsed_seqs_probs_mq20_clones.tsv --all_chains AB --nrepeat 1000 --tree_height_inches 5.0 --nbrdist_percentile 10
<br><br>
<u>Filename</u>:
./testing/test_small_mouse_pairseqs_v1_parsed_seqs_probs_mq20_clones_subject_trees.png
<br><br>
<u>Readme</u>:
This figure shows hierachical clustering (average linkage) trees of the different
mice (subjects), where the distance between two mice is defined as the average distance between TCRs in one
mouse and TCRs in the other. This could help us to interpret any high repertoire-heterogeneity Z-scores in the
"_subject_heterogeneity.png" bar plot earlier on, for example if it turns out that there are distinct sub-clusters of mice
with similar repertoires where the sub-clusters are quite different from one another.
See the earlier plot's readme for the explanation of how we assign a Z-score
and P-value to each repertoire reflecting the degree of repertoire heterogeneity across different mice. These
are the values shown after the epitope name above each row of trees. For this plot, more positive Z-scores mean
the intra-mouse distances are smaller than the inter-mouse distances (ie, more heterogeneity),
while negative Z-scores mean the opposite.
The first column of trees are the observed ones, while the next three are based on datasets in which the TCRs are
randomly reshuffled between mice (preserving each mouse's TCR count).
<br>
To provide a little more info about each mouse's individual repertoire, I looked at the intra-mouse distances for
TCRs from that mouse versus distances between randomly sampled sets of the same number of TCRs and assigned a Z-score
and P-value to each mouse (shown after the mouse names labeling each leaf in the trees). A negative Z-score means that
the mouse's TCRs are less diverse than average, while a positive Z-score means that the mouse's TCRs are more diverse
(more distant from one another). The P-value is a rough measure of how surprising these deviations from randomly
chosen TCRs actually are. The number of TCRs for each mouse is shown in parentheses after the mouse name. And *'s are
used to flag significant P-values (0.05, 0.01, 0.001).

<br><br>
<br>
<img src="test_small_mouse_pairseqs_v1_parsed_seqs_probs_mq20_clones_subject_trees.png" />
<br>
<br><br><h1>_vj_pairings.png</h1>

<u>Command</u>:
/shared/cs_researcher/Bradley_P/BradleyLab/pbradley/dat/tcr-dist/make_gene_plots.py --organism mouse --clones_file ./testing/test_small_mouse_pairseqs_v1_parsed_seqs_probs_mq20_clones.tsv --use_color_gradients
<br><br>
<u>Filename</u>:
./testing/test_small_mouse_pairseqs_v1_parsed_seqs_probs_mq20_clones_vj_pairings.png
<br><br>
<u>Readme</u>:
These diagrams depict the gene-segment pairing structure of the datasets. The four
genes are arrayed left to right with the alphas on the left and the betas on the right. Below each gene-type label (eg "VA")
is a color-stack showing all the TCR clones and how they break down into the different genes for that gene-type. Each clone
is devoted a constant vertical height in pixels indicated in the text at the top (N pixels in "Nx y-pixel scale"). The curved
segments joining neighboring gene-stacks show how the two gene distributions pair up, with the thickness of the segments
corresponding to the number of clones having those two segments (scaled by the indicated y-pixel scale). Significant gene-gene
pairings (positive or negative correlations with a P-value less than 1e-6) are labeled at the beginning and ending of the
corresponding segments. Gene-gene pairings which are not observed and for which this under-representation is significant
are indicated by dashed segments with P-value labels. Enrichments (depletions) of gene segments relative to
background are shown for all labeled genes by up (down) arrows where the number of arrowheads reflects the base-2
logarithm of the fold change, rounded down (one arrowhead means 2 <= fold change < 4,
two arrowheads means 4 <= fold change < 8, and so on).
<br>
<br>
The left-right ordering of the segment types is chosen so that VA and JA are on the left, VB and JB are on the right,
and the alpha-beta pairing with the largest adjusted mutual information is in the middle.

<br><br>
<br>
<img src="test_small_mouse_pairseqs_v1_parsed_seqs_probs_mq20_clones_vj_pairings.png" />
<br>
<br><br><h1>_motif_summary_ngbig.png</h1>

<u>Command</u>:
/shared/cs_researcher/Bradley_P/BradleyLab/pbradley/dat/tcr-dist/read_motifs.py --constant_seed --junction_bars --max_ng_lines 5000000 --organism mouse --clones_file ./testing/test_small_mouse_pairseqs_v1_parsed_seqs_probs_mq20_clones.tsv
<br><br>
<u>Filename</u>:
./testing/test_small_mouse_pairseqs_v1_parsed_seqs_probs_mq20_clones_motif_summary_ngbig.png
<br><br>
<u>Readme</u>:
This figure contains summary information on CDR3 motifs discovered by a simple
algorithm that basically looks for sequence patterns that occur more often in the epitope-specific datasets
than in a background set of sequences which is built by combining (1) random TCRs generated with a simple
model of the rearrangement process and (2) random subsets of next-gen unpaired data that is not selected for
epitope specificity. The challenge is to find motifs that are non-trivial in the sense that they do not just
come from germline sequence.
<br><br>
The main measure of significance I am using here is a simple "chi-squared" statistic which is basically <br>
<br>
 (observed-expected)*(observed-expected) / expected
<br><br>
Expected is judged based on two populations of background TCR sequences, one generated by a random model of the rearrangement process and one drawn from nextgen data. To be conservative, the value of expected used to calculate
significance is the maximum of the expected values based on these two background sets.
These background sets have the same V and J gene frequencies as the epitope specific set.
There is no correction for all the multiple testing going here, of course!
<br>
<br>
match- is the number of seed motif matches,<br>
match+ is the number of matched TCRs when I expand out to include nearby neighbors using the TCRdist distance measure
with a neighbor distance threshold of 19.26.
<br><br>
The match+ set are used to construct the sequence logos shown. The idea here is to account for the fact that the motif alphabet is not terribly expressive and motif discovery is not perfect, so we might be missing 'true' instances of the motif just using the initial seed motif (which is shown in "regular expression" syntax on the right):

<br><br>

^ = start of CDR3<br>
$ = end of CDR3<br>
. = any amino acid<br>
k = KR<br>
d = DE<br>
n = NQ<br>
s = ST<br>
f = FYWH<br>
a = AGSP<br>
v = VILM<br>
A = A<br>
C = C<br>
...<br>
Y = Y<br>
<br>
<br>
enrich is observed/expected
<br>
<br>
The color bars between the two logos indicate the inferred nucleotide source for the corresponding positions, as follows:<br>
<br>
light gray = V region<br>
black = D region<br>
dark gray = J region<br>
red = N insertions<br>
<br>

<br><br>
<br>
<img src="test_small_mouse_pairseqs_v1_parsed_seqs_probs_mq20_clones_motif_summary_ngbig.png" />
<br>
<br><br><h1>_tall_tree_AB.png</h1>

<u>Command</u>:
/shared/cs_researcher/Bradley_P/BradleyLab/pbradley/dat/tcr-dist/make_tall_trees.py --constant_seed --organism mouse --clones_file ./testing/test_small_mouse_pairseqs_v1_parsed_seqs_probs_mq20_clones.tsv --junction_bars
<br><br>
<u>Filename</u>:
./testing/test_small_mouse_pairseqs_v1_parsed_seqs_probs_mq20_clones_tall_tree_AB.png
<br><br>
<u>Readme</u>:

        These are TCRdist clustering trees for the different repertoires, with distances calculated over both chains. Down below this in the .html output are the clustering trees for distances calculated over each chain individually. To make these trees, the repertoire is first clustered with a fixed distance threshold (a TCRdist of 38.52 for single chain distances and 77.04 for paired alpha+beta chain distances) using a simple greedy approach that iteratively finds the TCR with the greatest number of neighbors within that distance, adds that cluster center and its neighbors to the list of clusters and deletes them from the repertoire, repeating until all the receptors have been clustered. These clusters are the leaves of the tree, and they are joined together for visualization purposes by an average-linkage hierarchical clustering approach that uses the matrix of distances between the cluster centers. The vertical thickness of the leaves and branches is proportional to the number of TCR clones represented by those branches. Repertoires with more than 300 receptors are subsampled to 300 after clustering so that the number of leaves in the tree doesn't become too large. Trees with all the TCR clones represented can be found by following the links toward the top of the .html output.<br><br>

        TCR logos are shown to the left of the tree for a representative subset of the branches (enclosed in dashed boxes labeled with their size; logos and boxed branches come in the same vertical order). Each logo panel shows the V- (left) and J- (right) gene frequencies in 'logo' format (height scaled by frequency, most frequent gene at the top; the IMGT gene names are trimmed to remove the leading TRAV/TRAJ/TRBV/TRBJ). In the middle is a CDR3 amino acid sequence logo. The colored bars below the CDR3 logo summarize the inferred rearrangement history of the grouped receptors by showing the nucleotide source, colored as follows: V region, light gray; J region, dark gray; D region, black; N insertions, red. THe number of TCRs contributing to the logo is shown to the left and should match the number next to the corresponding boxed branch of the tree.<br><br>

        This analysis is conducted at the level of clonotypes -- each expanded clone is condensed to a single receptor sequence for the purpose of clustering.

        
<br><br>
<br>
<img src="test_small_mouse_pairseqs_v1_parsed_seqs_probs_mq20_clones_tall_tree_AB.png" />
<br>
<br><br><h1>_tall_tree_A.png,_tall_tree_B.png</h1>
<img src="test_small_mouse_pairseqs_v1_parsed_seqs_probs_mq20_clones_tall_tree_A.png" />
<img src="test_small_mouse_pairseqs_v1_parsed_seqs_probs_mq20_clones_tall_tree_B.png" />
<br>
<br><br><h1>_F2_subject_tree.png</h1>

<u>Command</u>:
/shared/cs_researcher/Bradley_P/BradleyLab/pbradley/dat/tcr-dist/analyze_epitope_privacy.py --constant_seed --organism mouse --clones_file ./testing/test_small_mouse_pairseqs_v1_parsed_seqs_probs_mq20_clones.tsv --all_chains AB --nrepeat 1000 --tree_height_inches 5.0 --nbrdist_percentile 10
<br><br>
<u>Filename</u>:
./testing/test_small_mouse_pairseqs_v1_parsed_seqs_probs_mq20_clones_F2_subject_tree.png
<br><br>
<u>Readme</u>:

This is a hierarchical clustering dendrogram of the subjects for the F2 repertoire. The distance between a pair of subjects is defined to be the average distance between receptors in one subject and receptors in the other. The gene frequency composition of each subject is shown in the four stacks of colored bars in the middle (to the left of the tree, which is drawn with thin blue lines). To the right of the tree is a key for the gene segment coloring schemes which also shows the frequencies of each gene in the combined repertoire.

<br><br>
<br>
<img src="test_small_mouse_pairseqs_v1_parsed_seqs_probs_mq20_clones_F2_subject_tree.png" />
<br>
<br><br><h1>_M38_subject_tree.png</h1>

<u>Command</u>:
/shared/cs_researcher/Bradley_P/BradleyLab/pbradley/dat/tcr-dist/analyze_epitope_privacy.py --constant_seed --organism mouse --clones_file ./testing/test_small_mouse_pairseqs_v1_parsed_seqs_probs_mq20_clones.tsv --all_chains AB --nrepeat 1000 --tree_height_inches 5.0 --nbrdist_percentile 10
<br><br>
<u>Filename</u>:
./testing/test_small_mouse_pairseqs_v1_parsed_seqs_probs_mq20_clones_M38_subject_tree.png
<br><br>
<u>Readme</u>:

This is a hierarchical clustering dendrogram of the subjects for the M38 repertoire. The distance between a pair of subjects is defined to be the average distance between receptors in one subject and receptors in the other. The gene frequency composition of each subject is shown in the four stacks of colored bars in the middle (to the left of the tree, which is drawn with thin blue lines). To the right of the tree is a key for the gene segment coloring schemes which also shows the frequencies of each gene in the combined repertoire.

<br><br>
<br>
<img src="test_small_mouse_pairseqs_v1_parsed_seqs_probs_mq20_clones_M38_subject_tree.png" />
<br>
<br><br><h1>_M45_subject_tree.png</h1>

<u>Command</u>:
/shared/cs_researcher/Bradley_P/BradleyLab/pbradley/dat/tcr-dist/analyze_epitope_privacy.py --constant_seed --organism mouse --clones_file ./testing/test_small_mouse_pairseqs_v1_parsed_seqs_probs_mq20_clones.tsv --all_chains AB --nrepeat 1000 --tree_height_inches 5.0 --nbrdist_percentile 10
<br><br>
<u>Filename</u>:
./testing/test_small_mouse_pairseqs_v1_parsed_seqs_probs_mq20_clones_M45_subject_tree.png
<br><br>
<u>Readme</u>:

This is a hierarchical clustering dendrogram of the subjects for the M45 repertoire. The distance between a pair of subjects is defined to be the average distance between receptors in one subject and receptors in the other. The gene frequency composition of each subject is shown in the four stacks of colored bars in the middle (to the left of the tree, which is drawn with thin blue lines). To the right of the tree is a key for the gene segment coloring schemes which also shows the frequencies of each gene in the combined repertoire.

<br><br>
<br>
<img src="test_small_mouse_pairseqs_v1_parsed_seqs_probs_mq20_clones_M45_subject_tree.png" />
<br>

</body>
</html>
