{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "9ea4d77e-5c5b-4484-ac9f-bfc597a9c51e",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "0c681141-c98a-46bb-b1e6-394c4dfbe277",
   "metadata": {},
   "outputs": [],
   "source": [
    "dataset = pd.read_csv('vdjdb.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "27091e87-04dc-4970-9479-54b3a34aa627",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>complex.id</th>\n",
       "      <th>gene</th>\n",
       "      <th>cdr3</th>\n",
       "      <th>v.segm</th>\n",
       "      <th>j.segm</th>\n",
       "      <th>species</th>\n",
       "      <th>mhc.a</th>\n",
       "      <th>mhc.b</th>\n",
       "      <th>mhc.class</th>\n",
       "      <th>antigen.epitope</th>\n",
       "      <th>...</th>\n",
       "      <th>antigen.species</th>\n",
       "      <th>reference.id</th>\n",
       "      <th>method</th>\n",
       "      <th>meta</th>\n",
       "      <th>cdr3fix</th>\n",
       "      <th>vdjdb.score</th>\n",
       "      <th>web.method</th>\n",
       "      <th>web.method.seq</th>\n",
       "      <th>web.cdr3fix.nc</th>\n",
       "      <th>web.cdr3fix.unmp</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>TRA</td>\n",
       "      <td>CIVRAPGRADMRF</td>\n",
       "      <td>TRAV26-1*01</td>\n",
       "      <td>TRAJ43*01</td>\n",
       "      <td>HomoSapiens</td>\n",
       "      <td>HLA-B*08</td>\n",
       "      <td>B2M</td>\n",
       "      <td>MHCI</td>\n",
       "      <td>FLKEKGGL</td>\n",
       "      <td>...</td>\n",
       "      <td>HIV-1</td>\n",
       "      <td>PMID:15596521</td>\n",
       "      <td>{\"frequency\": \"\", \"identification\": \"tetramer-...</td>\n",
       "      <td>{\"cell.subset\": \"CD8+\", \"clone.id\": \"\", \"donor...</td>\n",
       "      <td>{\"cdr3\": \"CIVRAPGRADMRF\", \"cdr3_old\": \"CIVRAPG...</td>\n",
       "      <td>2</td>\n",
       "      <td>sort</td>\n",
       "      <td>sanger</td>\n",
       "      <td>no</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>TRB</td>\n",
       "      <td>CASSYLPGQGDHYSNQPQHF</td>\n",
       "      <td>TRBV13*01</td>\n",
       "      <td>TRBJ1-5*01</td>\n",
       "      <td>HomoSapiens</td>\n",
       "      <td>HLA-B*08</td>\n",
       "      <td>B2M</td>\n",
       "      <td>MHCI</td>\n",
       "      <td>FLKEKGGL</td>\n",
       "      <td>...</td>\n",
       "      <td>HIV-1</td>\n",
       "      <td>PMID:15596521</td>\n",
       "      <td>{\"frequency\": \"\", \"identification\": \"tetramer-...</td>\n",
       "      <td>{\"cell.subset\": \"CD8+\", \"clone.id\": \"\", \"donor...</td>\n",
       "      <td>{\"cdr3\": \"CASSYLPGQGDHYSNQPQHF\", \"cdr3_old\": \"...</td>\n",
       "      <td>2</td>\n",
       "      <td>sort</td>\n",
       "      <td>sanger</td>\n",
       "      <td>no</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0</td>\n",
       "      <td>TRB</td>\n",
       "      <td>CASSFEAGQGFFSNQPQHF</td>\n",
       "      <td>TRBV13*01</td>\n",
       "      <td>TRBJ1-5*01</td>\n",
       "      <td>HomoSapiens</td>\n",
       "      <td>HLA-B*08</td>\n",
       "      <td>B2M</td>\n",
       "      <td>MHCI</td>\n",
       "      <td>FLKEKGGL</td>\n",
       "      <td>...</td>\n",
       "      <td>HIV-1</td>\n",
       "      <td>PMID:15596521</td>\n",
       "      <td>{\"frequency\": \"\", \"identification\": \"tetramer-...</td>\n",
       "      <td>{\"cell.subset\": \"CD8+\", \"clone.id\": \"\", \"donor...</td>\n",
       "      <td>{\"cdr3\": \"CASSFEAGQGFFSNQPQHF\", \"cdr3_old\": \"C...</td>\n",
       "      <td>2</td>\n",
       "      <td>sort</td>\n",
       "      <td>sanger</td>\n",
       "      <td>no</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2</td>\n",
       "      <td>TRA</td>\n",
       "      <td>CAVPSGAGSYQLTF</td>\n",
       "      <td>TRAV20*01</td>\n",
       "      <td>TRAJ28*01</td>\n",
       "      <td>HomoSapiens</td>\n",
       "      <td>HLA-B*08</td>\n",
       "      <td>B2M</td>\n",
       "      <td>MHCI</td>\n",
       "      <td>FLKEKGGL</td>\n",
       "      <td>...</td>\n",
       "      <td>HIV-1</td>\n",
       "      <td>PMID:15596521</td>\n",
       "      <td>{\"frequency\": \"\", \"identification\": \"tetramer-...</td>\n",
       "      <td>{\"cell.subset\": \"CD8+\", \"clone.id\": \"\", \"donor...</td>\n",
       "      <td>{\"cdr3\": \"CAVPSGAGSYQLTF\", \"cdr3_old\": \"CAVPSG...</td>\n",
       "      <td>2</td>\n",
       "      <td>sort</td>\n",
       "      <td>sanger</td>\n",
       "      <td>no</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2</td>\n",
       "      <td>TRB</td>\n",
       "      <td>CASSFEPGQGFYSNQPQHF</td>\n",
       "      <td>TRBV13*01</td>\n",
       "      <td>TRBJ1-5*01</td>\n",
       "      <td>HomoSapiens</td>\n",
       "      <td>HLA-B*08</td>\n",
       "      <td>B2M</td>\n",
       "      <td>MHCI</td>\n",
       "      <td>FLKEKGGL</td>\n",
       "      <td>...</td>\n",
       "      <td>HIV-1</td>\n",
       "      <td>PMID:15596521</td>\n",
       "      <td>{\"frequency\": \"\", \"identification\": \"tetramer-...</td>\n",
       "      <td>{\"cell.subset\": \"CD8+\", \"clone.id\": \"\", \"donor...</td>\n",
       "      <td>{\"cdr3\": \"CASSFEPGQGFYSNQPQHF\", \"cdr3_old\": \"C...</td>\n",
       "      <td>2</td>\n",
       "      <td>sort</td>\n",
       "      <td>sanger</td>\n",
       "      <td>no</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   complex.id gene                  cdr3       v.segm      j.segm  \\\n",
       "0           1  TRA         CIVRAPGRADMRF  TRAV26-1*01   TRAJ43*01   \n",
       "1           1  TRB  CASSYLPGQGDHYSNQPQHF    TRBV13*01  TRBJ1-5*01   \n",
       "2           0  TRB   CASSFEAGQGFFSNQPQHF    TRBV13*01  TRBJ1-5*01   \n",
       "3           2  TRA        CAVPSGAGSYQLTF    TRAV20*01   TRAJ28*01   \n",
       "4           2  TRB   CASSFEPGQGFYSNQPQHF    TRBV13*01  TRBJ1-5*01   \n",
       "\n",
       "       species     mhc.a mhc.b mhc.class antigen.epitope  ... antigen.species  \\\n",
       "0  HomoSapiens  HLA-B*08   B2M      MHCI        FLKEKGGL  ...           HIV-1   \n",
       "1  HomoSapiens  HLA-B*08   B2M      MHCI        FLKEKGGL  ...           HIV-1   \n",
       "2  HomoSapiens  HLA-B*08   B2M      MHCI        FLKEKGGL  ...           HIV-1   \n",
       "3  HomoSapiens  HLA-B*08   B2M      MHCI        FLKEKGGL  ...           HIV-1   \n",
       "4  HomoSapiens  HLA-B*08   B2M      MHCI        FLKEKGGL  ...           HIV-1   \n",
       "\n",
       "    reference.id                                             method  \\\n",
       "0  PMID:15596521  {\"frequency\": \"\", \"identification\": \"tetramer-...   \n",
       "1  PMID:15596521  {\"frequency\": \"\", \"identification\": \"tetramer-...   \n",
       "2  PMID:15596521  {\"frequency\": \"\", \"identification\": \"tetramer-...   \n",
       "3  PMID:15596521  {\"frequency\": \"\", \"identification\": \"tetramer-...   \n",
       "4  PMID:15596521  {\"frequency\": \"\", \"identification\": \"tetramer-...   \n",
       "\n",
       "                                                meta  \\\n",
       "0  {\"cell.subset\": \"CD8+\", \"clone.id\": \"\", \"donor...   \n",
       "1  {\"cell.subset\": \"CD8+\", \"clone.id\": \"\", \"donor...   \n",
       "2  {\"cell.subset\": \"CD8+\", \"clone.id\": \"\", \"donor...   \n",
       "3  {\"cell.subset\": \"CD8+\", \"clone.id\": \"\", \"donor...   \n",
       "4  {\"cell.subset\": \"CD8+\", \"clone.id\": \"\", \"donor...   \n",
       "\n",
       "                                             cdr3fix vdjdb.score  web.method  \\\n",
       "0  {\"cdr3\": \"CIVRAPGRADMRF\", \"cdr3_old\": \"CIVRAPG...           2        sort   \n",
       "1  {\"cdr3\": \"CASSYLPGQGDHYSNQPQHF\", \"cdr3_old\": \"...           2        sort   \n",
       "2  {\"cdr3\": \"CASSFEAGQGFFSNQPQHF\", \"cdr3_old\": \"C...           2        sort   \n",
       "3  {\"cdr3\": \"CAVPSGAGSYQLTF\", \"cdr3_old\": \"CAVPSG...           2        sort   \n",
       "4  {\"cdr3\": \"CASSFEPGQGFYSNQPQHF\", \"cdr3_old\": \"C...           2        sort   \n",
       "\n",
       "  web.method.seq web.cdr3fix.nc web.cdr3fix.unmp  \n",
       "0         sanger             no               no  \n",
       "1         sanger             no               no  \n",
       "2         sanger             no               no  \n",
       "3         sanger             no               no  \n",
       "4         sanger             no               no  \n",
       "\n",
       "[5 rows x 21 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dataset.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "9c11c049-d87e-4eab-a647-33fe9ad6f1b0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "RangeIndex: 92771 entries, 0 to 92770\n",
      "Data columns (total 21 columns):\n",
      " #   Column            Non-Null Count  Dtype \n",
      "---  ------            --------------  ----- \n",
      " 0   complex.id        92771 non-null  int64 \n",
      " 1   gene              92771 non-null  object\n",
      " 2   cdr3              92771 non-null  object\n",
      " 3   v.segm            92670 non-null  object\n",
      " 4   j.segm            91626 non-null  object\n",
      " 5   species           92771 non-null  object\n",
      " 6   mhc.a             92771 non-null  object\n",
      " 7   mhc.b             92771 non-null  object\n",
      " 8   mhc.class         92771 non-null  object\n",
      " 9   antigen.epitope   92771 non-null  object\n",
      " 10  antigen.gene      92709 non-null  object\n",
      " 11  antigen.species   92771 non-null  object\n",
      " 12  reference.id      91260 non-null  object\n",
      " 13  method            92771 non-null  object\n",
      " 14  meta              92771 non-null  object\n",
      " 15  cdr3fix           92771 non-null  object\n",
      " 16  vdjdb.score       92771 non-null  int64 \n",
      " 17  web.method        92771 non-null  object\n",
      " 18  web.method.seq    92771 non-null  object\n",
      " 19  web.cdr3fix.nc    92771 non-null  object\n",
      " 20  web.cdr3fix.unmp  92771 non-null  object\n",
      "dtypes: int64(2), object(19)\n",
      "memory usage: 14.9+ MB\n"
     ]
    }
   ],
   "source": [
    "dataset.info()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e1d6cb90-e8cd-43ac-bfa7-b411bc3afb01",
   "metadata": {},
   "source": [
    "## Information about the dataset\n",
    "\n",
    "This is a a collapsed version of database used for annotation of single-chain TCR sequencing data by VDJdb-standalone software. <b>Each line corresponds to either TCR alpha or TCR beta record as specified by the gene column. TCR records coming from the same alpha-beta pair have the same index in complex.id column. In case complex.id is equal to 0 a record doesn't have either TCRalpha or TCRbeta chain information. </b> \n",
    "- Question: As stated above, if a particular row has complex.id = 0, then that record doesn't have either TCRalpha or beta chain information, then should we consider these rows for our model? Even though the record doesn't have any information about beta and alpha, it still has VDJ.score of 2. If the complex.id = 0, then why does the gene columns say TRB? What does it stands for?\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2e4399a3-160b-49d8-964d-2ba59dc4ef5d",
   "metadata": {},
   "source": [
    "### VDJ Score of rows with complex.id = 0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "6b831c99-580b-4137-a590-811c31c90510",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>vdjdb.score</th>\n",
       "      <th>count_with_complex_id = 0</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>25099</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>5169</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>1000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>315</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   vdjdb.score  count_with_complex_id = 0\n",
       "0            0                      25099\n",
       "1            1                       5169\n",
       "2            2                       1000\n",
       "3            3                        315"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dataset[dataset['complex.id'] == 0].groupby('vdjdb.score').size().to_frame('count_with_complex_id = 0').reset_index()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "790d2b64-55a8-4aea-8a85-e9a4318fcf5b",
   "metadata": {},
   "source": [
    "### Checking type of relationship between CDR3 and EPITOPE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "184d7774-6fc5-468b-b2c4-2c9777a5a9fe",
   "metadata": {},
   "outputs": [],
   "source": [
    "group_epitope = dataset.groupby('antigen.epitope')['cdr3'].unique().reset_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "1da44f3f-0c72-440c-b91c-c1264c02dc53",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>antigen.epitope</th>\n",
       "      <th>cdr3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AAFKRSCLK</td>\n",
       "      <td>[CVVNGNNNDMRF, CASSQGFGANVLTF, CVVNNNNNDMRF, C...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AAGIGILTV</td>\n",
       "      <td>[CAVNFGGGKLIF, CASSLSFGTEAFF, CAGGTGNQFYF, CAI...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AALALLLLDRLNQLE</td>\n",
       "      <td>[CALPREYGNKLVF, CASARRTSGEDTQYF, CAASWGSSNTGKL...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AALPILFQV</td>\n",
       "      <td>[CAGLPYGQNFVF, CASSPVGWGNTIYF]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AAVVRFQEAANKQKQ</td>\n",
       "      <td>[CAVSRAGAGSYQLTF, CSARAGGGEAKNIQYF, CAVRDPGNTD...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ADGLAYFRSSFKGG</td>\n",
       "      <td>[CAAEASNTNKVVF, CASSLNNANSDYTF]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ADLIAYLEQATKG</td>\n",
       "      <td>[CAAEPSSGQKLVF, CASSLNNANSDYTF]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ADLIAYLKQATKG</td>\n",
       "      <td>[CAAEPSSGQKLVF, CASSLNNANSDYTF, CAALRATGGNNKLT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>ADSLSFFSSSIKRGGGSLVP</td>\n",
       "      <td>[CAALRATGGNNKLTF, CASSLNWSQDTQYF]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>AEVQIDRLI</td>\n",
       "      <td>[CAVYDSNYQLIW, CASSEALSGGSETQYF, CASQGGQKLLF, ...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        antigen.epitope                                               cdr3\n",
       "0             AAFKRSCLK  [CVVNGNNNDMRF, CASSQGFGANVLTF, CVVNNNNNDMRF, C...\n",
       "1             AAGIGILTV  [CAVNFGGGKLIF, CASSLSFGTEAFF, CAGGTGNQFYF, CAI...\n",
       "2       AALALLLLDRLNQLE  [CALPREYGNKLVF, CASARRTSGEDTQYF, CAASWGSSNTGKL...\n",
       "3             AALPILFQV                     [CAGLPYGQNFVF, CASSPVGWGNTIYF]\n",
       "4       AAVVRFQEAANKQKQ  [CAVSRAGAGSYQLTF, CSARAGGGEAKNIQYF, CAVRDPGNTD...\n",
       "5        ADGLAYFRSSFKGG                    [CAAEASNTNKVVF, CASSLNNANSDYTF]\n",
       "6         ADLIAYLEQATKG                    [CAAEPSSGQKLVF, CASSLNNANSDYTF]\n",
       "7         ADLIAYLKQATKG  [CAAEPSSGQKLVF, CASSLNNANSDYTF, CAALRATGGNNKLT...\n",
       "8  ADSLSFFSSSIKRGGGSLVP                  [CAALRATGGNNKLTF, CASSLNWSQDTQYF]\n",
       "9             AEVQIDRLI  [CAVYDSNYQLIW, CASSEALSGGSETQYF, CASQGGQKLLF, ..."
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "group_epitope.head(10)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8f96faea-b344-48b3-9d09-71c9c4d556bc",
   "metadata": {},
   "source": [
    "### Inference\n",
    "\n",
    "There is a 1 (epitope) to many(cdr3) relation, this means that a single epitope can be cured by many variants of CDR3 combination."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8b27d6b8-8953-449e-8329-3cf22e8be686",
   "metadata": {},
   "source": [
    "### Find number of records available for each vdj score"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "ec6ed746-eec6-40d2-8bd9-8fa8aef4e77d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>vdjdb.score</th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>80357</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>8329</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>2336</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>1749</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   vdjdb.score  count\n",
       "0            0  80357\n",
       "1            1   8329\n",
       "2            2   2336\n",
       "3            3   1749"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dataset.groupby('vdjdb.score').size().to_frame('count').reset_index()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "886f8a99-6709-4a9c-8374-e47426cb0ac9",
   "metadata": {},
   "source": [
    "## Inference \n",
    "\n",
    "According to README-Metadata (https://github.com/UoB-DSMP-2023-24/dsmp-2024-group-5/blob/Nakul/README-MetaData.md#vdjdb-scoring)\n",
    "\n",
    "The TCR:peptide:MHC complexes are assigned with confidence scores (VDJdb Scoring) by evaluating TCR sequence, identification and verification confidence on the following criteria:\n",
    "        \n",
    "        1. Ensuring TCR sequence is correctly identified according to method.sequencing and method.singlecell (1-3 points)\n",
    "        2. Initial identification of TCR:pMHC is correct according to method.identification (0-1 point)\n",
    "        3. Verification T-cell specificity (0-3 points)\n",
    "        \n",
    "- The final score is then calculated as minimal between score from part 1. and sum of scores from part 2. and part 3..\n",
    "score: description \n",
    "\n",
    "VDJ score | description\n",
    "------|----------------------\n",
    "0     | Low confidence/no information - a critical aspect of sequencing/specificity validation is missing\n",
    "1     | Moderate confidence - no verification / poor TCR sequence confidence\n",
    "2     | High confidence - has some specificity verification, good TCR sequence confidence\n",
    "3     | Very high confidence - has extensive verification or structural data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "18ad0edc-dc4c-41b0-95b6-5af7266484bc",
   "metadata": {},
   "source": [
    "### Possible columns to be taken for further analysis\n",
    "\n",
    "1. Cdr3\n",
    "2. Peptide\n",
    "3. mhc.a\n",
    "4. mhc.b\n",
    "5. complex.id\n",
    "6. vdjdb.score\n",
    "7. v.segment\n",
    "8. j.segment"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6d5e6e7e-3a3e-4cad-b6f6-4067a8b91abc",
   "metadata": {},
   "source": [
    "---"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "182a1cd7-7328-4987-a032-47200c37725d",
   "metadata": {},
   "source": [
    "## Visualisations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "66289036-4f42-4109-9b82-26510ea55be3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Correlation matrix of above columns\n",
    "target_columns = ['antigen.epitope']\n",
    "feature_columns = ['complex.id', 'gene', 'cdr3', 'v.segm', 'j.segm','mhc.a','mhc.b']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "98316541-56a2-49b6-8786-a5f4c733580b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['complex.id', 'gene', 'cdr3', 'v.segm', 'j.segm', 'species', 'mhc.a',\n",
       "       'mhc.b', 'mhc.class', 'antigen.epitope', 'antigen.gene',\n",
       "       'antigen.species', 'reference.id', 'method', 'meta', 'cdr3fix',\n",
       "       'vdjdb.score', 'web.method', 'web.method.seq', 'web.cdr3fix.nc',\n",
       "       'web.cdr3fix.unmp'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Pairplot for feature-target relationships\n",
    "sns.pairplot(dataset, x_vars=['feature_1', 'feature_2', 'feature_3'], y_vars=[target_columns], height=4, aspect=1.2)\n",
    "plt.suptitle('Pairplot of Features and Target', y=1.02)\n",
    "plt.show()\n",
    "\n",
    "# Correlation heatmap between features and target\n",
    "correlation_matrix = df.corr()\n",
    "plt.figure(figsize=(10, 8))\n",
    "sns.heatmap(correlation_matrix[['target']], annot=True, cmap='coolwarm', fmt='.2f', linewidths=.5)\n",
    "plt.title('Correlation between Features and Target')\n",
    "plt.show()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
